vmware-event-router/internal/processor/knative/knative.go

```diff
@@ -0,0 +1,238 @@
+	}
+
+	target := uri.String()
+	client, err := ceClient(target, cfg.InsecureSSL, cfg.Encoding)
+	if err != nil {
+		return nil, err
+	}
+
+	p.Logger = kLog
+	p.ceClient = client
+	p.sink = target
+	p.stats = metrics.EventStats{
+		Provider:    string(config.ProcessorKnative),
+		Type:        config.EventProcessor,
+		Address:     p.sink,
+		Started:     time.Now().UTC(),
+		Invocations: make(map[string]*metrics.InvocationDetails),
+	}
+
+	go p.PushMetrics(ctx, ms)
+	return &p, nil
+}
+
+// ceClient returns a cloud events client for the given URI and TLS insecure value
+func ceClient(uri string, insecure bool, encoding string) (ceclient.Client, error) {
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: insecure, //nolint:gosec
+	}
+	httpTransport := &http.Transport{TLSClientConfig: &tlsConfig}
+
+	// Create protocol and client
+	transport, err := cloudevents.NewHTTP(cloudevents.WithTarget(uri), cloudevents.WithRoundTripper(httpTransport))
+	if err != nil {
+		return nil, errors.Wrap(err, "create cloud events http transport")
+	}
+
+	clientOpts := []ceclient.Option{ceclient.WithUUIDs()}
+	switch encoding {
+	case "structured":
+		clientOpts = append(clientOpts, ceclient.WithForceStructured())
+	case "binary":
+		clientOpts = append(clientOpts, ceclient.WithForceBinary())
+	default:
+		return nil, fmt.Errorf("unsupported encoding type specified: %q", encoding)
+	}
+
+	client, err := cloudevents.NewClient(transport, clientOpts...)
+	if err != nil {
+		return nil, errors.Wrap(err, "create cloud events client")
+	}
+
+	return client, nil
+}
+
+// Process processes a cloud event. Internal retry logic handles transient
+// processing errors. If an event cannot be processed an error is returned.
+func (p *Processor) Process(ctx context.Context, ce cloudevents.Event) error {
+	if p.isStopped() {
+		return ErrStopped
+	}
+
+	// coordinate concurrent shutdown
+	p.wg.Add(1)
+	defer p.wg.Done()
+
+	p.Debugw("processing event", "eventID", ce.ID(), "event", ce)
+	subject := ce.Subject()
+	p.mu.Lock()
+	// initialize invocation stats
+	if _, ok := p.stats.Invocations[subject]; !ok {
+		p.stats.Invocations[subject] = &metrics.InvocationDetails{}
+	}
+	p.mu.Unlock()
+
+	// register retry options
+	ctx = cloudevents.ContextWithRetriesExponentialBackoff(ctx, retryDelay, maxRetries)
+	p.Infow("sending event", "eventID", ce.ID(), "subject", subject)
+	result := p.ceClient.Send(ctx, ce)
+	p.Debugw("got response", "eventID", ce.ID(), "response", result)
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if !cloudevents.IsACK(result) {
+		p.stats.Invocations[subject].Failure()
+		return processor.NewError(config.ProcessorKnative, errors.Wrapf(result, "send event %s", ce.ID()))
+	}
+
+	p.Infow("successfully sent event", "eventID", ce.ID())
+	p.stats.Invocations[subject].Success()
+	return nil
+}
+
+func (p *Processor) PushMetrics(ctx context.Context, ms metrics.Receiver) {
+	ticker := time.NewTicker(metrics.PushInterval)
+	defer ticker.Stop()
+
```

Comment:
Should we check whether processor is stopped here? Or we do that only for process and shutdown?

Reply:
`PushMetrics` will return on context cancelled

Reply:
Yeah I see now... I should've had more attention to detail